StandardLibraryEssentials/StandardTemplateLibrary/UnorderedContainers/task.md

```diff
@@ -0,0 +1,96 @@
+
+The standard library provides the `std::hash` function for basic types, and you can define your own hash function for any custom types.
+
+There are a few functions for buckets manipulation:
+* `bucket_count()` - returns the number of buckets
+* `bucket_size(n)` - returns the number of elements in the bucket `n`
+* `bucket(k)` - returns the bucket number for the key `k`
+* `load_factor()` - returns the average number of elements per bucket
+* `max_load_factor()` - returns the maximum load factor
+* `rehash(n)` - sets the number of buckets to `n` or more
+
+When you need a custom hash function, you can define it like this:
+```cpp
+struct my_type {
+private:
+    int x;
+    int y;
+
+public:
+    my_type(const int x, const int y) : x(x), y(y) {};
+    
+    bool operator==(const my_type& other) const {
+        return x == other.x && y == other.y;
+    }
+
+    friend struct my_hash;
+};
+
+struct my_hash {
+    std::size_t operator()(const my_type& t) const {
+        return std::hash<int>()(t.x) ^ std::hash<int>()(t.y);
+    }
+};
+
+int main() {
+    std::unordered_set<my_type, my_hash> s;
+    s.emplace(1, 2);
+}
+```
+
+There are multi-versions of these containers: `std::unordered_multiset`, `std::unordered_multimap`. They allow storing multiple elements with the same key.
+
+As an example, let's take a look at the following code:
+```cpp
+#include <iostream>
+#include <unordered_set>
+#include <string>
+
+int main() {
+    std::unordered_set<std::string> set;
+    set.insert("Greetings, traveler!");
+    set.insert("How are you?");
+    set.insert("Come join us!");
+    
+    if (set.find("How are you?") != set.end()) {
+        std::cout << "Found!" << std::endl;
+    }
+    
+    // If you're using C++20 or above, you can use this:
+    
+    // if (set.contains("How are you?")) {
+    //     std::cout << "Found!" << std::endl;
+    // }
+    
+    // print all elements, order is not guaranteed
+    for (const auto& s : set) {
+        std::cout << s << std::endl;
+    }
+}
+```
+
+You need to implement a [cache](https://en.wikipedia.org/wiki/Cache_(computing)) system with an [LRU](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU)) (Least Recently Used) policy. Its member functions are defined in the `/include/Cache.h` file. Since cache is a very fast type of storage, you will need to use hashing to implement it. The provided solution uses `std::list` and `std::unordered_map`, but you can choose any container from the STL for your own implementation. At first, define a structure inside `/include/Cache.h` file that will store key-value pairs. Then, define another structure that will store the key and an iterator to the corresponding element in the list. After defining these structures, you can implement the cache operations on `task.cpp`.
```

Comment:
> The provided solution uses `std::list` and `std::unordered_map`, but you can choose any container from the STL for your own implementation.

Maybe let's remove this suggestion? I think it can only confuse the students. 
Using just `std::list` and `std::unordered_map` is fine for this task.